### PR TITLE
[NemotronH] Do not force router to run in fp32

### PR DESCRIPTION
**Pull request by roikoren755**
## Purpose
Current code forces the MoE router computation to FP32, even though checkpoints have it in bfloat16. This takes up about 40% of the forward pass, under normal workloads, and does not provide an accuracy boost.

This PR removes this limitations.

## Test Plan
No additional tests, all tests pass, accuracy does not degrade

## Test Result
All tests pass, accuracy did not degrade

Running GSM8K, got the following results.

PR:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5671|±  |0.0136|
|     |       |strict-match    |     5|exact_match|↑  |0.8431|±  |0.0100|
```

Main:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5572|±  |0.0137|
|     |       |strict-match    |     5|exact_match|↑  |0.8453|±  |0.0100|
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/flashinfer_trtllm_moe.py [hunk 1]
```diff
@@ -309,6 +309,14 @@ def fi_trtllm_fp8_per_tensor_moe(
 
     from vllm.utils.flashinfer import flashinfer_trtllm_fp8_per_tensor_scale_moe
 
+    # The DeepSeekV3 routing method requires float32 router logits.
+    print(routing_method_type)
+    if routing_method_type == RoutingMethodType.DeepSeekV3:
+        routing_logits = routing_logits.to(torch.float32)
+
+    if routing_bias is not None:
+        routing_bias = routing_bias.to(hidden_states.dtype)
```
Comment (tomeras91) [inline vllm/model_executor/layers/fused_moe/flashinfer_trtllm_moe.py]: I remember something about it being important that the bias is in FP32.. I understand in this case we first cast logits to FP32 (since we're using DS routing) so the bias is actually in FP32, but doesn't it make more sense to cast logits to bias dtype instead of the other way around?

**Review by mgoin — APPROVED** [review]
LGTM, nice find! We maybe should add an assert on the bias, but this roughly matches other trtllm moe impls. Do you have any perf result? You mentioned this takes 40% of time, which doesn't make sense to me
